File: Go/logger.go
package main

import (
	"fmt"
	"os"
	"sync"
	"time"
)

// Logger

type Logger struct {
	mu   sync.Mutex
	file *os.File
}

func NewLogger(filename string) *Logger {
	file, err := os.Create(filename)
	if err != nil {
		fmt.Fprintf(os.Stderr, "ERROR: Could not create log file: %v\n", err)
		return &Logger{}
	}
	l := &Logger{file: file}
	l.Log("LOGGER", "Log file initialized: "+filename)
	return l
}

func (l *Logger) Log(source, message string) {
	l.mu.Lock()
	defer l.mu.Unlock()

	timestamp := time.Now().Format("15:04:05.000")
	logLine := fmt.Sprintf("[%s] [%-12s] %s", timestamp, source, message)
	fmt.Println(logLine)

	if l.file != nil {
		fmt.Fprintln(l.file, logLine)
	}
}

func (l *Logger) Close() {
	l.mu.Lock()
	defer l.mu.Unlock()

	if l.file != nil {
		l.file.Close()
		l.file = nil
	}
}

// Write results to file with error handling using defer

func writeResultsToFile(filename string, results []string) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("could not create file: %w", err)
	}
	defer file.Close() // Ensure file is closed with defer

	fmt.Fprintln(file, "= Processing Results")

	fmt.Fprintln(file)

	for _, result := range results {
		fmt.Fprintln(file, result)
	}

	fmt.Fprintln(file)
	fmt.Fprintf(file, "Total Rides Processed: %d\n", len(results))

	return nil
}

File: Java/Logger.java
package ridesharing;

import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.locks.ReentrantLock;

class Logger {
    private static final ReentrantLock lock = new ReentrantLock();
    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("HH:mm:ss.SSS");
    private static PrintWriter fileWriter = null;

    public static void init(String filename) {
        try {
            fileWriter = new PrintWriter(new FileWriter(filename, false));
            log("LOGGER", "Log file initialized: " + filename);
        } catch (IOException e) {
            System.err.println("ERROR: Could not create log file: " + e.getMessage());
        }
    }

    public static void log(String source, String message) {
        lock.lock();
        try {
            String timestamp = LocalDateTime.now().format(formatter);
            String threadName = Thread.currentThread().getName();
            String logLine = String.format("[%s] [%-10s] [%-12s] %s", timestamp, threadName, source, message);
            System.out.println(logLine);
            if (fileWriter != null) {
                fileWriter.println(logLine);
                fileWriter.flush();
            }
        } finally {
            lock.unlock();
        }
    }

    public static void close() {
        lock.lock();
        try {
            if (fileWriter != null) {
                fileWriter.close();
                fileWriter = null;
            }
        } finally {
            lock.unlock();
        }
    }
}